Automattic/woocommerce-payments
# Enables deferred intent UPE for existing legacy card stores

Closes https://github.com/Automattic/woocommerce-payments/issues/7221

❗ **Should be merged between Oct 9 and Oct 27**. Keeping the PR status _Draft_ to avoid an accidental merge.

#### Changes proposed in this Pull Request
This PR is the third and the last step of migration to deferred intent creation UPE. Previously, we migrated split UPE stores in https://github.com/Automattic/woocommerce-payments/pull/7058 and legacy UPE stores in https://github.com/Automattic/woocommerce-payments/pull/7306. During legacy UPE stores migration, we switched to the server-side feature flag, which made the implementation in https://github.com/Automattic/woocommerce-payments/pull/7306 a bit more complex than it could have been because we needed to support both client-side and server-side stores. This PR removes all this complexity since, starting at `6.7.0`, each store should rely only on the server-side feature flag.









#### Testing instructions





Speaking high-level, we need to check that legacy card/legacy UPE/split UPE stores are now all using deferred intent creation UPE, that there's no fallback available for merchants but if we turn off the dUPE toggle for legacy card or legacy UPE, the store should roll back with corresponding notice in the Payments settings. 

> Pre-requisite: In your local server code, change https://github.com/Automattic/woocommerce-payments-server/pull/3963/files#diff-c79fe5913d8e8943eac11e58823d53e42ce3e6234736f0ed0c745fa901b53c17R1844 to compare against 6.5.0 so that your local is treated as eligible, for testing purposes

##### Legacy card stores

1. For [this configuration](https://github.com/Automattic/woocommerce-payments/assets/22319444/40d458b9-571c-4e17-a320-77d6cd3ec2c1) (no UPE is enabled), confirm that `'is_deferred_intent_creation_upe_enabled' => true,` in the account cache contents. This means that by default, even if the feature flag doesn't exist in the server's DB, deferred UPE will be treated as enabled by the server. 
2. Navigate to `Payments -> Settings` and confirm that your settings [look the following](https://github.com/Automattic/woocommerce-payments/assets/22319444/568c869f-f2b1-4d16-b26e-d251f9fa7f42), meaning no notices, banners, but UPE methods
3. Click the kebab menu on the upper right and confirm that only the _Provide feedback_ button is available, but not _Disable_, like it was previously.
4. Perform a test purchase, navigate to Stripe Dashboard, and confirm that in Payment's details, `gateway_type` equals `split_upe_with_deferred_intent_creation` in the metadata section.

##### Legacy UPE stores
* For [this configuration](https://github.com/Automattic/woocommerce-payments/assets/22319444/0344199a-2245-4c92-8ff3-392a62baf749), confirm that steps 2-4 from the previous paragraph are leading to the same result.
* In your local server environment, run `wp wcpay disable_dupe acct_{YOUR_ACCOUNT_ID}`. With this action, we're testing the process of disabling dUPE for merchants if needed. Refresh the cache afterward.
* Navigate to payments settings and confirm that your settings [look the following](https://github.com/Automattic/woocommerce-payments/assets/22319444/d042c04a-0813-4eee-bc1a-dadfc387ed49). The text is being currently aligned with the Design Team and I'll update the text with the final version once confirmed. 
* Perform a purchase, go to Stripe Dashboard to confirm `gateway_type = legacy_upe` in the payment's details.
* Enable deferred UPE flag back with `wp wcpay enable_dupe acct_{YOUR_ACCOUNT_ID}` and refresh the cache
* Navigate to _Payments -> Settings_ and confirm that the notice shown previously disappeared. This means your store now uses deferred UPE again.

##### Split UPE stores
* For [this configuration](https://github.com/Automattic/woocommerce-payments/assets/22319444/f98b17c9-4780-419f-847c-4e352d0bea2c), confirm that steps 2-4 from the previous paragraph are leading to the same result.
* We're not testing split UPE roll back because split UPE stores were migrated some time ago, and no roll back for them is considered.


-------------------


**Post merge**
